service/src/scripts/approve_utils.ts

```diff
@@ -0,0 +1,61 @@
+import 'dotenv/config'
+import { Web3Interface } from './web3-interface'
+import { ethers } from 'ethers'
+import { exit } from 'process'
+import { RPC_EXCEPTION } from '../utils/exit-codes'
+import { archLogger } from '../utils/chalk-theme'
+import createpromt from 'prompt-sync';
+
+const prompt = createpromt({ sigint: true });
+
+const _hasAllowance = async (web3Interface: Web3Interface) => {
+  const allowance = await web3Interface.sarcoToken.allowance(web3Interface.ethWallet.address, process.env.SARCO_DIAMOND_ADDRESS!);
+
+  if (allowance.eq(ethers.constants.MaxUint256)) {
```

Comment:
https://etherscan.io/token/0x7697b462a7c4ff5f8b55bdbc2f4076c2af9cf51a#code#L422

During a `transferFrom` call on SARCO, the approved amount decreases by the amount being transferred.

So, if a user had previously approved `MaxUint256`, then this code here would "work" the first time through (correctly notice that no approval transaction needs to be made), but it would "fail" the next time this arch service attempts to call a sarcophagus contract function that moves SARCO tokens (require another approval transaction), despite there still being a stupidly high amount of approved tokens (just slightly below `MaxUint256`.

I don't know the context of when this function is called. Maybe just once upon service startup? If so, then I submit that the code here will _work_, but it's a bit of a brute force solution. Approval transactions will happen way more than necessary.

My suggestion would be, that when attempting to determine if an approval transaction needs to be made, that should always be done with the context of the amount of SARCO that are going to be transferred.

For example, this function could take a second parameter, which is compared against the actual allowance amount

```ts
const _hasAllowance = async (
  web3Interface: Web3Interface,
  sarcoToTransfer: BigNumber
) => {
const allowance = await web3Interface.sarcoToken.allowance(web3Interface.ethWallet.address, process.env.SARCO_DIAMOND_ADDRESS!);

  if (allowance.gte(sarcoToTransfer)) {
archLogger.notice("Already approved");
    return true;
  }

  return false;
}
```

Of course, this might require some upstream dependencies, like making sure that every time a Sarcophagus transaction (that results is SARCO being `transferFrom`d) is about to be signed and broadcasted, this "allowance check" is done.

Reply:
Oh riiight.

Thanks for calling this out, whew. I'll push an update in a bit.